### PR TITLE
WIP - reworking some sliver persistent header logic

### PR DESCRIPTION
**Pull request by Piinks**
Part of https://github.com/flutter/flutter/issues/126448

This is only for the pinned header, not floating or pinned-floating (I have other potential changes for those).

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/material/app_bar.dart [hunk 1]
```diff
@@ -1254,7 +1254,7 @@ class _SliverAppBarDelegate extends SliverPersistentHeaderDelegate {
     final double extraToolbarHeight = math.max(minExtent - _bottomHeight - topPadding - (toolbarHeight ?? kToolbarHeight), 0.0);
     final double visibleToolbarHeight = visibleMainHeight - _bottomHeight - extraToolbarHeight;
 
-    final bool isScrolledUnder = overlapsContent || forceElevated || (pinned && shrinkOffset > maxExtent - minExtent);
+    final bool isScrolledUnder = overlapsContent || forceElevated;
```
Comment (Piinks) [inline packages/flutter/lib/src/material/app_bar.dart:1257]: Looks like at some point we put in this work around @HansMuller

**Comment by Piinks** [comment]
@HansMuller will you give this a try with the custom header you have been writing? I know some tests fail, I have to check them out, but I think this fixes the pinned case (floating-pinned and floating not included here)

**Comment by HansMuller** [comment]
This change resolves the problem I found with creating a SliverPersistent header using the code in https://gist.github.com/HansMuller/480b464ca764096be8cf775270977617. I tried some variations, like putting slivers above the header and making the viewport smaller than the header. Everything worked as expected.

**Comment by Piinks** [comment]
Thanks for checking!

**Comment by Piinks** [comment]
Found a case here where this is wrong, overlap should remain false until the min extent is reached in the case where maxExtent != minExtent. Will keep investigating.

**Comment by Piinks** [comment]
Ok, I think this is correct for all cases now. I'll need to make a migration plan for affected customers.
@thkim1011 do you want to try this in your main axis SliverGroup? This may be why there was some odd behavior. This only fixes when the persistent header is pinned. Still need to check floating, and floating+pinned.

**Comment by Piinks** [comment]
Actually I found another case this does not agree with. Hold off on testing.

**Comment by Hixie** [comment]
Closing per @Piinks

**Pull request closed without merging**